### PR TITLE
ENH Support sparse X for glmnet and explicit behavior wrt to non standardized y

### PR DESCRIPTION
**Pull request by mathurinm**

### CODE REVIEW

**Comment by mathurinm** [comment]
@jolars when l1_ratio is not 1, it seems glmnet does not converge to the correct limit. Do you see something suspicious in the code ?

image

image

**Comment by jolars** [comment]
Uh, disregard the last post, I wasn't looking in the right place.

**Comment by jolars** [comment]
Hm, I cannot find anything that would explain this right now. Does it only happen when `fit_intercept=True`?

**Comment by mathurinm** [comment]
fit intercept=False does the same: 
image

**Comment by mathurinm** [comment]
As @Klopfe told me, glmnet centers and scales the target.

Doing it manually in the data generation process fixes stuff: 
image


But removing it and passing `"standardize.response": False` to glmnet does not fix the issue...

**Comment by jolars** [comment]
I think standardize.response is just for the multivariate lasso. standardize decides if the X matrix is centered and scaled. y is always centered if the intercept is fit I believe, but none of that should matter since glmnet returns unstandardized coefficients.
________________________________

**Comment by Klopfe** [comment]
yes, it does not make sense to me but it was already seen before (https://stats.stackexchange.com/questions/206898/difference-between-elasticnet-in-scikit-learn-python-and-glmnet-in-r). It's very weird... After liblinear doing weird things on the SVM solver, it seems that glmnet does some weird things as well.

**Comment by jolars** [comment]
I'm e-mailing Trevor Hastie now. Let's see what he says.

File: objective.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ def __init__(self, l1_ratio=.5, reg=1.0, fit_intercept=False):
     def set_data(self, X, y):
         self.X, self.y = X, y
         self.n_samples, self.n_features = self.X.shape
-        self.lmbda_max = np.max(np.abs(X.T @ y)) / len(y)
+        self.lmbda_max = np.max(np.abs(X.T @ y)) / (len(y) * self.l1_ratio)
```
Comment (jolars) [inline objective.py]: Shouldn't `lambda_max` be adjusted if the intercept is fit as in https://github.com/benchopt/benchmark_lasso/blob/ff69e6b0bf9ef8e172346000fabf87489de84b9f/objective.py#L40-L44 ?

Comment (Klopfe) [reply objective.py]: yes, you are right.

Comment (mathurinm) [reply objective.py]: thanks @jolars

**Comment by mathurinm** [comment]
Trevor Hastie's answer:

> In the documentation we warn that for gaussian family glmnet standardizes y before computing the path (and then unstandardizes afterward)
> With elastic net this is not equivariant. So I suggest you standardize y (with unit standardization using 1/n variance formula) and see if the problem persists 

So we have our reason. Indeed on news20 which is almost standardized the problem was less visible.

**Pull request merged**